### PR TITLE
[WIP] cuDNN Legacy MultiHeadAttention Integration

### PR DESCRIPTION
**Pull request by mingxu1067**
### PR types

New features
### PR changes

OPs
### Describe

- Added two new ops, `mha` and `mha_data_prepare`.
> mha: compute multi-head attention.
> mha_data_prepare: convert attention mask to `mha` op's required meta data.
- Added cuDNN implementaion to `MultiHeadAttn` layer
> No code changes need.
> Enable cuDNN implementation when compiled with cuda and `cudnn_version` >= 8300
> Compatible with original implementaion on save/load behaviors.

### CODE REVIEW

File: paddle/fluid/operators/mha/mha_op.cc [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+        .AsDispensable();
+
+    AddOutput("output",
+              "(Tensor) The output tensor of MultiHeadAttention (MHA) op.");
+
+    AddAttr<std::string>(
+        "cache_key",
+        "(string) A key to obtain space for saving cuDNN descriptors");
+    AddAttr<bool>(
+        "is_training",
```
Comment (Xreki) [inline paddle/fluid/operators/mha/mha_op.cc:204]: 用`is_test`

File: paddle/fluid/platform/dynload/cudnn.cc [hunk 2]
```diff
@@ -43,6 +44,10 @@ CUDNN_DNN_ROUTINE_EACH_AFTER_R7(DEFINE_WRAP);
 CUDNN_DNN_ROUTINE_EACH_R8(DEFINE_WRAP);
 #endif
 
+#ifdef CUDNN_DNN_ROUTINE_EACH_AFTER_TWO_R8
```
Comment (Xreki) [inline paddle/fluid/platform/dynload/cudnn.cc:47]: 不用修改老的这一处代码了。

File: paddle/fluid/operators/CMakeLists.txt [hunk 3]
```diff
@@ -26,6 +26,11 @@ add_subdirectory(prim_ops)
 if(WITH_MKLDNN)
     add_subdirectory(mkldnn)
 endif()
+if (WITH_GPU)
+    if (${CUDNN_VERSION} VERSION_GREATER_EQUAL 8300)
```
Comment (Xreki) [inline paddle/fluid/operators/CMakeLists.txt:30]: - dynload代码里面写的`CUDA_VERSION >= 8200`，这里写的`8300`，版本不一致？
- 另外，我觉得没有必要新建一个子文件夹，新增算子放在`fused`目录？

File: paddle/fluid/operators/mha/mha_data_prepare_op.h [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+  void Compute(const framework::ExecutionContext& context) const override {
+    platform::Place host_pinned_place = platform::CUDAPinnedPlace();
+
+    const Tensor* attn_mask = context.Input<Tensor>("attn_mask");
+    // TODO(Ming Huang): Use reduce_sum kernel to compute qkvo seqlen in GPU
+    // buffer. That could remove this DtoH data movement.
+    size_t attn_mask_size = attn_mask->numel() * sizeof(int);
+    memory::allocation::AllocationPtr attn_mask_host =
+        memory::Alloc(host_pinned_place, attn_mask_size);
+    PADDLE_ENFORCE_GPU_SUCCESS(
```
Comment (Xreki) [inline paddle/fluid/operators/mha/mha_data_prepare_op.h:39]: 可以用`memory::Copy`吗？

File: paddle/fluid/operators/mha/mha_data_prepare_op.h [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+    int batch = attn_mask->dims()[0];
+    int seqlen = attn_mask->dims()[3];
+    int stride = phi::product(attn_mask->dims()) / batch;
+
+    int* attn_mask_host_ptr = reinterpret_cast<int*>(attn_mask_host->ptr());
+    for (int i = 0; i < batch; ++i) {
+      int* begin = attn_mask_host_ptr + stride * i;
+      int* end = begin + seqlen;
+      qo_kv_seqlen_host_ptr[i] = std::accumulate(begin, end, 0);
```
Comment (Xreki) [inline paddle/fluid/operators/mha/mha_data_prepare_op.h:59]: 数据从GPU -> Pinned memory，这里使用，需要加sync吗？

File: paddle/fluid/operators/mha/mha_op.cc [hunk 6]
```diff
@@ -0,0 +1,347 @@
...
+    return framework::OpKernelType(expected_kernel_type.data_type_,
+                                   expected_kernel_type.place_,
+                                   tensor.layout());
+  }
+};
+
+class MHAOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  void Make() override {
+    AddInput(
```
Comment (Xreki) [inline paddle/fluid/operators/mha/mha_op.cc:152]: 输入、输出变量名，用`XxxXxx`这种大小写混合的驼峰式命名。

File: paddle/fluid/operators/mha/mha_op.cc [hunk 7]
```diff
@@ -0,0 +1,347 @@
...
+        "query",
+        "(Tensor) The input tensor query of MultiHeadAttention (MHA) op."
+        " The dimension of query should be 3 (batch, seqlen, embedding_size).");
+    AddInput(
+        "key",
+        "(Tensor) The input tensor key of MultiHeadAttention (MHA) op."
+        " The dimension of key should be 3 (batch, seqlen, embedding_size).");
+    AddInput(
+        "value",
```
Comment (Xreki) [inline paddle/fluid/operators/mha/mha_op.cc:161]: `key`和`value`，若和`query`相同，应该允许不设置。

File: paddle/fluid/operators/mha/mha_op.cc [hunk 8]
```diff
@@ -0,0 +1,347 @@
...
+              "(Tensor) The output tensor of MultiHeadAttention (MHA) op.");
+
+    AddAttr<std::string>(
+        "cache_key",
+        "(string) A key to obtain space for saving cuDNN descriptors");
+    AddAttr<bool>(
+        "is_training",
+        "(bool) Indicate training forward or inference. Default is true.")
+        .SetDefault(true);
+    AddAttr<bool>(
```
Comment (Xreki) [inline paddle/fluid/operators/mha/mha_op.cc:207]: attention中有4个bias，只能统一设置有或没有吗？

File: paddle/fluid/operators/mha/mha_op.cu [hunk 9]
```diff
@@ -0,0 +1,29 @@
+/* Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
```
Comment (Xreki) [inline paddle/fluid/operators/mha/mha_op.cu:1]: 若无CUDA Kernel，文件后缀可用`.cu.cc`

File: paddle/fluid/operators/mha/mha_op.h [hunk 10]
```diff
@@ -0,0 +1,505 @@
...
+    int beam_size = 1;  // not support beam_dim currently.
+
+    // TODO(Ming Huang): Work around now is using user-defined key
+    const std::string cache_key = context.Attr<std::string>("cache_key");
+
+    const Tensor* qo_kv_seqlen = context.Input<Tensor>("qo_kv_seqlen");
+    const int* qo_kv_seqlen_data = qo_kv_seqlen->data<int>();
+    const Tensor* qo_kv_seqlen_host =
+        context.Input<Tensor>("qo_kv_seqlen_host");
+    int* qo_kv_seqlen_data_host;
```
Comment (Xreki) [inline paddle/fluid/operators/mha/mha_op.h:132]: 一些为调cudnn接口进行的准备工作，可以封装成函数，使得Kernel流程看起来简单、清晰些。

File: paddle/fluid/operators/mha/mha_op.h [hunk 11]
```diff
@@ -0,0 +1,505 @@
...
+    }
+
+    auto dtype = platform::CudnnDataType<T>::type;
+    auto comp_prec =
+        dtype == CUDNN_DATA_DOUBLE ? CUDNN_DATA_DOUBLE : CUDNN_DATA_FLOAT;
+
+    auto cudnn_handle = dev_ctx.cudnn_handle();
+
+    // Setup Pre Attention Dropout
+    cudnnDropoutDescriptor_t pre_dropout_desc = nullptr;
```
Comment (Xreki) [inline paddle/fluid/operators/mha/mha_op.h:202]: dropout desc不用cache起来吗？

**Review by Xreki — COMMENTED** [review]
建议可以拆分下pr，逐渐合入。

**Comment by jeng1220** [comment]
@mingxu1067 ,
Please close this PR.

**Comment by jeng1220** [comment]
We are moving forward to cuDNN frontend. The cuDNN 8.9 will provide **much** better fusion kernels for MHA. Thus, close this PR.
